homeassistant/components/pglab/discovery.py

```diff
@@ -0,0 +1,284 @@
+        entity_registry = er.async_get(hass)
+        for platform, unique_id in discovery_info.entities:
+            if entity_id := entity_registry.async_get_entity_id(
+                platform, DOMAIN, unique_id
+            ):
+                entity_registry.async_remove(entity_id)
+
+        # Destroy the device.
+        device_registry = dr.async_get(hass)
+        if device_entry := device_registry.async_get_device(
+            identifiers={(DOMAIN, device_id)}
+        ):
+            device_registry.async_remove_device(device_entry.id)
+
+        # Clean the discovery info.
+        del self._discovered[device_id]
+
+    async def start(
+        self, hass: HomeAssistant, mqtt: PyPGLabMqttClient, entry: PGLABConfigEntry
+    ) -> None:
+        """Start discovering a PGLab devices."""
+
+        async def discovery_message_received(msg: ReceiveMessage) -> None:
+            """Received a new discovery message."""
+
+            # Create a PGLab device and add entities.
+            try:
+                pglab_device = await self.__build_device(mqtt, msg)
+            except PGLabDiscoveryError as err:
+                _LOGGER.info("Can't create PGLabDiscovery instance(%s) ", str(err))
+
+                # For some reason it's not possible to create the device with the discovery message,
+                # be sure that any previous device with the same topic is now destroyed.
+                device_id = get_device_id_from_discovery_topic(msg.topic)
+
+                # If there is a valid topic device_id clean everything relative to the device.
+                if device_id:
+                    self.__clean_discovered_device(hass, device_id)
+
+                return
+
+            # Create a new device.
+            device_registry = dr.async_get(hass)
+            device_registry.async_get_or_create(
+                config_entry_id=entry.entry_id,
+                configuration_url=f"http://{pglab_device.ip}/",
+                connections={(CONNECTION_NETWORK_MAC, pglab_device.mac)},
+                identifiers={(DOMAIN, pglab_device.id)},
+                manufacturer=pglab_device.manufactor,
+                model=pglab_device.type,
+                name=pglab_device.name,
+                sw_version=pglab_device.firmware_version,
+                hw_version=pglab_device.hardware_version,
+            )
+
+            # Do some checking if previous entities must be updated.
+            if pglab_device.id in self._discovered:
+                # The device is already been discovered,
+                # get the old discovery info data.
+                discovery_info = self._discovered[pglab_device.id]
+
+                if discovery_info.hash == pglab_device.hash:
+                    # Best case, there is nothing to do.
+                    # The device is still in the same configuration. Same name, same shutters, same relay etc.
+                    return None
+
+                _LOGGER.info(
```

Comment:
Please don't log on info

Reply:
About the first point... yes I agree with you. Entity must have a valid Unique_id, otherwise there is a problem somewhere else.  I'm going to raise an exception if the unique_id is None.  About the second point... I'll log  on warning.